### ISSUE TITLE
Typescript stops type checking when using Object.assign()

### ISSUE THREAD

**Issue by imdavidmin** [microsoft/TypeScript#45521]
# Bug Report

### 🔎 Search Terms

React Dispatch SetStateAction Object

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about all

### ⏯ Playground Link

```ts
import React, { useState, Dispatch, SetStateAction } from 'react'

type someObj = {
  a: string,
  b: string
}

const x: someObj = { a: 'a', b: 'b' }

const [y, setY]: [someObj, Dispatch<SetStateAction<someObj>>] = useState(x)

function pushIntoY(value: any) {
  setY(Object.assign({}, y, { c: value }))
}

pushIntoY(1)
```

<details><summary><b>Output</b></summary>

```ts
import { useState } from 'react';
const x = { a: 'a', b: 'b' };
const [y, setY] = useState(x);
function pushIntoY(value) {
    setY(Object.assign({}, y, { c: value }));
}
pushIntoY(1);

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "strict": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "target": "ES2017",
    "jsx": "react",
    "module": "ESNext",
    "moduleResolution": "node"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDGMA0cDecCuAzkgMowoxLYAiwhYFaAFtiUjGRUgIIbAQA7OAF84AMygQQcAORRUGGQCglMAJ5gkcQlKQB5AEYArOAF5cSuHBQAubTCjABAc0yW4Bu4QdPnS4Spogt5wAB5euoYm5ni2sigy2J6yBjIigcHwANpq2MQwAJoAunZZOiD6xjR0DDDMADxsHOSUvDD8AvXllUYAfL1FZgTEnJQAFKEAlCpi+AJ8gnBgREwAkgIwEAVjAG4oADb4SHYoAmqTFlb521FIGAB0KISEwM4CYzjC2Lm4cGh2e0OWmEk2mASUy0Iaw2WzGAEZJkA)
      
### 💻 Code

```
import React, { useState, Dispatch, SetStateAction } from 'react'

type someObj = {
  a: string,
  b: string
}

const x: someObj = { a: 'a', b: 'b' }

const [y, setY]: [someObj, Dispatch<SetStateAction<someObj>>] = useState(x)

function pushIntoY(value: any) {
  setY(Object.assign({}, y, { c: 3 }))
}
```

### 🙁 Actual behavior

Can push anything into something that is supposed to be strictly typed, with no error prompt in IDE.

### 🙂 Expected behavior

{c: 3} should show an error saying 'c' is not a key of someObj type.

**Comment by RyanCavanaugh** [comment]
Object types aren't sealed; it is not a type error to have an excess key. In other scenarios where we can detect an immediate issue, excess property checking *can* flag these, but this is intentionally not one of those situations.

**Comment by fatcerberus** [comment]
Hmm... shouldn't *this* be an error, though?

```ts
import React, { useState, Dispatch, SetStateAction } from 'react'

type someObj = {
  a: string,
  b: string,
  c: string,
}

const x: someObj = { a: 'a', b: 'b', c: 'c' }

const [y, setY]: [someObj, Dispatch<SetStateAction<someObj>>] = useState(x)

function pushIntoY(value: number) {
  setY(Object.assign({}, y, { c: value }))
}

pushIntoY(1)
```

`string & number` is an empty type so this should flag an error somewhere around the `setY` call, but doesn't.

**Comment by imdavidmin** [comment]
Updated the code sample to below

```
import React, { useState, Dispatch, SetStateAction } from 'react'

type someObj = {
  a: string,
  b: string
}

const x: someObj = { a: 'a', b: 'b' }
const x2: someObj = {a: 'a', b: 'b', c: 1}

const [y, setY]: [someObj, Dispatch<SetStateAction<someObj>>] = useState(x)

function pushIntoY(value: any) {
  setY(Object.assign({}, y, { c: value }))
}

pushIntoY(1)
```

I can't understand the thinking behind making the statement declaring x2 result in an error, but for y does not? In both cases the end result is the same. 

As @fatcerberus also shows in his example - The setY call in IDE clearly shows its a `Dispatch<SetStateAction<someObj>`, so `Object.assign()` statement there should result in a `someObj` type return value?

**Comment by fatcerberus** [comment]
`x2` is an error for a different reason: excess property checking, which is more of a linter check than a proper type check.  @RyanCavanaugh is correct that, in general, object types allow excess properties; this is structural subtyping.  EPC in fresh object literals is primarily done to catch typos.

I realized after I posted that your example and mine are different: In my example, `c` exists on the type and I've assigned an incompatible value to it due to an unsoundness hole w.r.t. type intersection.  In yours, `c` is just an extra property, which is allowed per structural subtyping as being returned from a function makes it no longer subject to EPC.

That being said, the `setY()` call *is* subject to EPC if you change the `Object.assign` call to an equivalent spread:
```ts
setY({ ...y, c: value });  // error, like x2
```

https://www.typescriptlang.org/docs/handbook/type-compatibility.html#starting-out

> Note that `dog` has an extra `owner` property, but this does not create an error. Only members of the target type (`Pet` in this case) are considered when checking for compatibility.
